File: backend/app/schemas/rentals.py
"""Rental, location, and incident request/response schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class LocationBrief(BaseModel):
    id: UUID
    name: str
    address: str | None = None
    city: str | None = None
    status: str
    lat: float | None = None
    lng: float | None = None
    available_devices: int = 0

    model_config = {"from_attributes": True}


class LocationDetail(LocationBrief):
    working_hours: dict | None = None
    contacts: dict | None = None
    photo_url: str | None = None
    capacity: int = 0


class DeviceBrief(BaseModel):
    id: UUID
    model: str
    short_code: str
    color: str | None = None
    storage: str
    condition_grade: int | None = None

    model_config = {"from_attributes": True}


class RentalCreateRequest(BaseModel):
    device_id: UUID
    location_id: UUID
    with_udobno_subscription: bool = False


class RentalBrief(BaseModel):
    id: UUID
    status: str
    device: DeviceBrief | None = None
    location_name: str | None = None
    activated_at: datetime | None = None
    paid_until: datetime | None = None
    next_charge_at: datetime | None = None
    deposit_amount: float | None = None
    total_charged: float = 0
    debt_amount: float = 0
    booking_expires_at: datetime | None = None

    model_config = {"from_attributes": True}


class RentalDetail(RentalBrief):
    created_at: datetime | None = None
    closed_at: datetime | None = None
    has_udobno_at_booking: bool | None = None


class RentalListResponse(BaseModel):
    items: list[RentalBrief]
    total: int


class ConfirmPickupResponse(BaseModel):
    status: str
    activated_at: datetime
    paid_until: datetime


class IncidentCreateRequest(BaseModel):
    type: str
    description: str
    photo_urls: list[str] = []


class IncidentBrief(BaseModel):
    id: UUID
    type: str
    status: str
    description: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: apps/client/src/api/rentals.ts
import { api } from "./client";

export type RentalBrief = {
  id: string;
  status: string;
  device: { id: string; model: string; short_code: string; color: string | null };
  location_name: string;
  activated_at: string | null;
  paid_until: string | null;
  next_charge_at: string | null;
  deposit_amount: number | null;
  total_charged: number;
  debt_amount: number;
  booking_expires_at: string | null;
};

export type RentalListResponse = {
  items: RentalBrief[];
  total: number;
};

export type IncidentBrief = {
  id: string;
  type: string;
  status: string;
  description: string | null;
  created_at: string;
};

export const rentalsApi = {
  create: (data: { device_id: string; location_id: string; with_udobno_subscription?: boolean }) =>
    api.post<RentalBrief>("/client/rentals", data),

  list: (filter: "active" | "history" = "active") =>
    api.get<RentalListResponse>("/client/rentals", { status_filter: filter }),

  get: (id: string) => api.get<RentalBrief>(`/client/rentals/${id}`),

  cancelBooking: (id: string) =>
    api.post(`/client/rentals/${id}/cancel-booking`),

  confirmPickup: (id: string) =>
    api.post(`/client/rentals/${id}/confirm-pickup`),

  reportPickupIssue: (id: string, data: { issue_type: string; description: string; photo_urls?: string[] }) =>
    api.post(`/client/rentals/${id}/report-pickup-issue`, data),

  initReturn: (id: string) =>
    api.post(`/client/rentals/${id}/init-return`),

  confirmReturn: (id: string, data: { return_session_id: string }) =>
    api.post(`/client/rentals/${id}/confirm-return`, data),

  reportReturnDispute: (id: string) =>
    api.post(`/client/rentals/${id}/report-return-dispute`),

  createIncident: (data: { rental_id: string; type: string; description: string; photo_urls?: string[] }) =>
    api.post<IncidentBrief>("/client/incidents", data),

  listIncidents: () => api.get<IncidentBrief[]>("/client/incidents"),
};
